demo/popcha-client/web/src/authentication.py

```diff
@@ -0,0 +1,99 @@
+"""
+This file includes all the authentication logic. No external server logic is
+done here.
+"""
+from __future__ import annotations
+
+import secrets
+import time
+from urllib import parse
+
+import jwt
+from jwt import PyJWTError
+from werkzeug.datastructures import MultiDict
+
+
+class Authentication:
+    """
+    Class that allows users to authenticate
+    """
+
+    def __init__(self):
+        self.providers: list = []
+        self.login_states: dict[(str, str, float)] = {}
+
+    def get_url(self, auth_server:str, lao_id: str, host_server: str, 
+                host_port: str | int, client_id: str) -> str:
+        """
+        Generates the url to contact the authentication server. It contains all
+        the information required by the authentication server
+        :param auth_server: The auth server domain
+        :param lao_id: The LAO ID the user wants to connect with
+        :param host_server: The current host server
+        :param host_port: The port the host server is running on
+        :param client_id: The unique identifier of this client
+        :return: A url to the authentication server
+        """
+        nonce = secrets.token_urlsafe(64)
+        state = secrets.token_urlsafe(64)
+        self.login_states[state] = (nonce, auth_server, time.time())
+        parameters = {
+            "response_mode": "query",
+            "response_type": "id_token",
+            "client_id": client_id,
+            "redirect_uri": f"https://{host_server}:{str(host_port)}/cb",
+            "scope": "openid profile",
+            "login_hint": lao_id,
+            "nonce": nonce,
+            "state": state
+            }
+        return f"https://{auth_server}/authorize?{parse.urlencode(parameters)}"
+
+    def validate_args(self, args: MultiDict[str, str], client_id: str) \
+            -> str | None:
+        """
+        Validate the args as if they were coming from OpenID Authentication request
+        :param args: A MultiDict containing the arguments (almost the same as a
+        dict).
+        :param client_id: The ID of this client
+        :return: The unique user identifier if logged in None otherwise
+        """
+
+        missing_arg: bool = ("id_token" not in args or "token_type" not in
+                             args or "state" not in args)
+        if missing_arg:
+            return None
+
+        if args.get("state") not in self.login_states.keys():
+            return
+        nonce_data = self.login_states.pop(args.get("state"))
```

Comment:
Yes, all of your concerns should be fixed at a later stage. Since it is only an example server and the token creation time is already checked while the token is decoded, I do not think this is that important. In addition, an attacker could easily add new providers to this website through the `/add_provider` get request.